File: oxidize-python/oxidize_python/core/tokenizer/tiktoken.py
from __future__ import annotations

import re
from dataclasses import dataclass, field

from oxidize_python.core.tokenizer.tokenizer import EncodeOptions, SpecialTokens


@dataclass
class TiktokenTokenizer:
    pattern: str = ""
    ranks: dict[str, float] = field(default_factory=dict)
    reverse: dict[int, str] = field(default_factory=dict)
    special: SpecialTokens = field(default_factory=SpecialTokens)
    _compiled: re.Pattern[str] | None = None

    def __post_init__(self) -> None:
        if not self.pattern:
            self.pattern = r"'s|'t|'re|'ve|'m|'ll|'d| ?[A-Za-z]+| ?[0-9]+| ?[^\sA-Za-z0-9]+|\s+"
        self._compiled = re.compile(self.pattern)

    def name(self) -> str:
        return "tiktoken"

    def special_tokens(self) -> SpecialTokens:
        return self.special

    def vocab_size(self) -> int:
        return len(self.ranks)

    def encode(self, text: str, opts: EncodeOptions) -> list[int]:
        if not text:
            return []
        assert self._compiled is not None
        out: list[int] = []
        for chunk in self._compiled.findall(text):
            out.extend(self._bpe(chunk.encode()))
        return out

    def _bpe(self, word: bytes) -> list[int]:
        if len(word) == 1:
            return [self._lookup(word.decode())]
        parts = [word.decode()]
        while True:
            pairs = [(parts[i], parts[i + 1]) for i in range(len(parts) - 1)]
            best_idx = -1
            best_rank = 1e30
            for i, p in enumerate(pairs):
                key = p[0] + p[1]
                if key in self.ranks:
                    rank = self.ranks[key]
                    if best_idx < 0 or rank < best_rank:
                        best_idx, best_rank = i, rank
            if best_idx < 0:
                break
            parts = (
                parts[:best_idx]
                + [parts[best_idx] + parts[best_idx + 1]]
                + parts[best_idx + 2 :]
            )
        return [self._lookup(p) for p in parts]

    def _lookup(self, p: str) -> int:
        return int(self.ranks.get(p, 0))

    def decode(self, tokens: list[int]) -> str:
        sb = []
        for tid in tokens:
            if tid in (self.special.bos, self.special.eos, self.special.pad):
                continue
            piece = self.reverse.get(tid)
            if piece:
                sb.append(piece)
            else:
                sb.append(chr(tid))
        return "".join(sb)

    def decode_skip_special(self, tokens: list[int]) -> str:
        return self.decode(tokens)

File: oxidize-golang/core/tokenizer/tiktoken.go
package tokenizer

import (
	"regexp"
	"strings"
)

// TiktokenTokenizer implements a regex-based BPE tokenizer compatible with
// OpenAI's tiktoken cl100k_base / o200k_base encodings.
type TiktokenTokenizer struct {
	pattern *regexp.Regexp
	ranks   map[string]float32
	reverse map[uint32]string
	special SpecialTokens
}

// NewTiktokenTokenizer constructs a Tiktoken tokenizer with a regex pattern
// and a map of merge-pair -> rank.
func NewTiktokenTokenizer(pattern string, ranks map[string]float32, special SpecialTokens) *TiktokenTokenizer {
	if pattern == "" {
		pattern = defaultTiktokenPattern()
	}
	return &TiktokenTokenizer{
		pattern: regexp.MustCompile(pattern),
		ranks:   ranks,
		reverse: map[uint32]string{},
		special: special,
	}
}

func (t *TiktokenTokenizer) Name() string { return "tiktoken" }
func (t *TiktokenTokenizer) SpecialTokens() SpecialTokens { return t.special }
func (t *TiktokenTokenizer) VocabSize() int               { return len(t.ranks) }

func (t *TiktokenTokenizer) Encode(text string, opts EncodeOptions) ([]uint32, error) {
	if text == "" {
		return []uint32{}, nil
	}
	chunks := t.pattern.FindAllString(text, -1)
	out := make([]uint32, 0, len(chunks))
	for _, chunk := range chunks {
		ids := t.bpe([]byte(chunk))
		out = append(out, ids...)
	}
	return out, nil
}

func (t *TiktokenTokenizer) bpe(word []byte) []uint32 {
	if len(word) == 1 {
		return []uint32{t.lookup(string(word))}
	}
	parts := []string{string(word)}
	for {
		pairs := make([][2]string, 0, len(parts)-1)
		for i := 0; i < len(parts)-1; i++ {
			pairs = append(pairs, [2]string{parts[i], parts[i+1]})
		}
		bestIdx := -1
		bestRank := float32(1e30)
		for i, p := range pairs {
			if rank, ok := t.ranks[p[0]+p[1]]; ok {
				if bestIdx < 0 || rank < bestRank {
					bestIdx = i
					bestRank = rank
				}
			}
		}
		if bestIdx < 0 {
			break
		}
		merged := make([]string, 0, len(parts)-1)
		merged = append(merged, parts[:bestIdx]...)
		merged = append(merged, parts[bestIdx]+parts[bestIdx+1])
		merged = append(merged, parts[bestIdx+2:]...)
		parts = merged
	}
	out := make([]uint32, 0, len(parts))
	for _, p := range parts {
		out = append(out, t.lookup(p))
	}
	return out
}

func (t *TiktokenTokenizer) lookup(p string) uint32 {
	if t.ranks == nil {
		return 0
	}
	if rank, ok := t.ranks[p]; ok {
		return uint32(rank)
	}
	return 0
}

func (t *TiktokenTokenizer) Decode(tokens []uint32) (string, error) {
	var sb strings.Builder
	for _, id := range tokens {
		if id == t.special.BOS || id == t.special.EOS || id == t.special.Pad {
			continue
		}
		if piece, ok := t.reverse[id]; ok {
			sb.WriteString(piece)
		} else {
			sb.WriteRune(rune(id))
		}
	}
	return sb.String(), nil
}

func (t *TiktokenTokenizer) DecodeSkipSpecial(tokens []uint32) (string, error) {
	return t.Decode(tokens)
}
